drivers/nec/display.cr

```diff
@@ -0,0 +1,308 @@
+      end
+    end
+
+    task.try &.success
+  end
+
+  private def parse_response(data : String)
+    # 14..15 == type (we don't care)
+    value = data[20..23].to_i(16)
+    command = Command.from_value(data[10..13].to_i(16))
+
+    logger.debug { "command is 0x#{data[10..13]}" }
+    logger.debug { command }
+    logger.debug { "value is 0x#{data[20..23]}" }
+    logger.debug { value }
+
+    case command
+    when .video_input?
+      self[:input] = Input.from_value(value)
+    when .audio_input?
+      self[:audio] = Audio.from_value(value)
+    when .volume_status?
+      self[:volume] = value
+      self[:audio_mute] = value == 0
+    when .brightness_status?
+      self[:brightness] = value
+    when .contrast_status?
+      self[:contrast] = value
+    when .mute_status?
+      self[:audio_mute] = value == 1
+      self[:volume] = 0 if value == 1
+    when .auto_setup?
+      # auto_setup
+      # nothing needed to do here (we are delaying the next command by 4 seconds)
+    else
+      logger.info { "-- NEC LCD, unknown response: #{data[10..13]}" }
+      logger.info { "-- NEC LCD, full response was: #{data}" }
+    end
+  end
+
+  # Types of messages sent to and from the LCD
+  enum MsgType : UInt8
+    Command           = 0x41 # 'A'
+    CommandReply      = 0x42 # 'B'
+    GetParameter      = 0x43 # 'C'
+    GetParameterReply = 0x44 # 'D'
+    SetParameter      = 0x45 # 'E'
+    SetParameterReply = 0x46 # 'F'
+  end
+
+  enum Command
+    VideoInput       = 0x0060
+    AudioInput       = 0x022E
+    VolumeStatus     = 0x0062
+    MuteStatus       = 0x008D
+    PowerOnDelay     = 0x02D8
+    ContrastStatus   = 0x0012
+    BrightnessStatus = 0x0010
+    AutoSetup        = 0x001E
+    PowerQuery       = 0x01D6
+    Save             = 0x0C
+  end
+
+  private def format_value(value : Int, length : Int = 4) : String
+    length = 2 if value == 0x0C # To deal with Command::Save
+    value.to_s(16, true).rjust(length, '0')
+  end
+
+  {% for name in Command.constants %}
+    @[Security(Level::Administrator)]
+    def {{name.id.underscore}}(priority : Int32 = 0)
+      do_send(MsgType::GetParameter, Command::{{name.id}}, priority: priority, name: {{name.id.underscore.stringify}})
+    end
+  {% end %}
+
+  private def check_checksum(data : Bytes)
+    checksum = 0x00_u8
+    # Loop through the second to the third last element
+    if data.size >= 2
+      data[1..-3].each do |b|
+        checksum = checksum ^ b
+      end
+      # Check the checksum equals the second last element
+      logger.debug { "Error: checksum should be #{checksum.to_s(16)}" } unless checksum == data[-2]
+      checksum == data[-2]
+    else
+      true
+    end
+  end
+
+  # Builds the command and creates the checksum
+  private def do_send(type : MsgType, data : Command | Array(Int), **options)
+    data = [data.value] if data.is_a?(Command)
+    data = data.join { |i| format_value(i) }.bytes
+    bytes = Bytes.new(data.size + 11)
```

Comment:
When it comes to building the packet here, take a look at [String.build](https://crystal-lang.org/api/0.35.1/String.html#build(capacity=64,&):self-class-method). There's a tiny bit of overhead here that just using `Bytes` directly, but this may be counteracted by the improved readability.

This will make concat'ing multiple byte values, and some of the type conversion a little clear:
```crystal
String.build do |str|
  str.write_byte 0x01 # write bytes directly
  str << "hello"      # or strings
  str << 'c'          # and char values too
  str << 1            # integers will write as their direct string form
end
```

Reply:
Thanks @KimBurgess , I've made this pretty big commit to address your latest 2 comments.

So I think the protocol is general enough so that all commands will be in this general format:
`#{msg_type}#{command}#{data}`

`data` is not required for all commands so is only passed in when necessary which is for any commands with `MsgType::SetParameter` and in the only exception for `def power` for commands with `MsgType::Command`

Reply:
Also, I couldn't figure out how to generate the checksum with one `String.build` block annoyingly so I had to settle on doing this.

https://github.com/PlaceOS/drivers/blob/10db226249ede7583b834fb73a52b839a4f5004e/drivers/nec/display.cr#L285-L301

Not sure if it's possible to generate the entire command string in one `String.build` block unless you know some way to do this @KimBurgess

Reply:
Oh just realised I had to use `String.build` 3 times actually :/

Reply:
Nice. Heaps neater.